SQL FlattenSchema Rewrite (#587)
SimpleDataLabsInc/prophecy-docs: docs/analysts/development/gems/visual-expression-builder/variant-schema.md

Prophecy does not store variant data types within the table definition. Each row can vary in data types, which makes them difficult to infer and use. Fortunately, you don't have to infer the schema yourself. You can use the column selector inside of your gems to automatically infer the variant data type, explore the multi-type variant structure, and later select a nested column to use in your transformations.

To automatically infer the variant data type:

1. Open a gem that uses a variant column input, such as the [FlattenSchema gem](/analysts/flatten-schema).

1. Click the **Variant** dropdown, and click **Infer Schema**.

   Prophecy automatically detects and identifies the variant data types in your input data.

   ![Schema and column selector](img/variant-infer-schema.png)

   :::note

   Prophecy caches the inferred schema so you can use it again in the future whenever you reopen the model, gem, or another gem connected to the same input port.

   To see the last time your variant data type was inferred, see the box at the bottom of the column selector.

   To refresh the schema, click **Infer Schema** again.

   :::

### Edit the variant data type schema

> If Prophecy did not cover your schema case while sampling the records, you can make edits yourself by:

Reviewer: ```suggestion
If Prophecy missed certain schema cases while sampling the records, you can make edits yourself by:
```
basically emphasizing that the sample was not big enough - it didn't sample a record that included a different schema

> 1. Click the **Variant** dropdown, and click **Edit Schema**.

Reviewer: fyi "you can make edits yourself by" ... clicking rather than click? I'll let you decide how to change this

Author: I ended up breaking it up -- I made a new clause after on doing it.

1. Use the data type dropdowns to manually choose the data type of each nested schema.

   ![Edit schema view](img/variant-edit-schema.png)

## Variant sample setting

When Prophecy infers the variant data type, it samples the records to identify all potential iterations of keys and values within the schema. The default number of records that Prophecy parses to understand the nested data schema is 100.

To update this limit:

> 1. Click on `...` at the top of the page.

Reviewer: I've been bolding `...` but I like this better... maybe we can change everywhere eventually

1. Click **Development Settings**.
1. Put in the number of records you want Prophecy to parse to understand the nested data schema.

   We recommend that you increase the limit for small structures, or decrease it for larger ones.

   ![Variant sampling setting](img/variant-sampling-setting.png)

:::note
This setting does not rely on the ratio of the data since that would require a complete count of the data records.
:::

> ## Add a nested column

Reviewer: Maybe a clearer title would be "use a nested column in an expression" ? When I read this title (which I understand is old), I'm like, add a nested column where?
